### ISSUE TITLE
HLS m3u8 - audio but no video on Android

### ISSUE THREAD

**Issue by nlbailey** [media-kit/media-kit#701]
I am implementing media-kit into a project and I'm running into an issue with the playback of a .m3u8 file. It works fine on iOS, but on Android I am only getting audio.

Here is a link to a file for testing: http://isqmedia.com/HLS/nginx.m3u8

The test file is only maybe 15 seconds long. Video should display a purple (pink?) box and the words "This is a test" moving around. Audio is a music track that will fade out / in once during playback. Just to clarify expected behavior of the test file.

This is what I get when accessing in app:
D/roup.audio_tes( 2310): PlayerBase::PlayerBase()
D/roup.audio_tes( 2310): TrackPlayerBase::TrackPlayerBase()
I/libOpenSLES( 2310): Emulating old channel mask behavior (ignoring positional mask 0x3, using default mask 0x3 based on channel count of 2)
D/AudioTrack( 2310): createTrack_l(0): AUDIO_OUTPUT_FLAG_FAST denied by server; frameCount 0 -> 2132
E/libOpenSLES( 2310): Configuration error: unknown key
W/libOpenSLES( 2310): Leaving AndroidConfiguration::GetConfiguration (SL_RESULT_PARAMETER_INVALID)
D/roup.audio_tes( 2310): PlayerBase::stop() from IPlayer
D/AudioTrack( 2310): stop(32): called with 0 frames delivered

I'm assuming it has to do with the configuration error?

**Comment by nlbailey** [comment]
I set it to false but still no video with audio. Here is the output when playing now:

I/flutter ( 6804): media_kit: wakelock: _count = 1
I/media_kit( 6804): com.alexmercerind.media_kit_video.VideoOutputManager.create: 2312054896
I/media_kit( 6804): flutterJNIAPIAvailable = true
I/media_kit( 6804): com.alexmercerind.media_kit_video.VideoOutput: id = 7
I/flutter ( 6804): {id: 7}
I/MediaCodec( 6804): (0x7a962000) init name(OMX.qcom.video.decoder.avc)
I/OMXClient( 6804): IOmx service obtained
I/MediaCodec( 6804): (0x7a962000) Component Allocated (OMX.qcom.video.decoder.avc)
I/MediaCodec( 6804): (0x7a962000) configure surface(0x0) crypto(0x0) flags(0)
D/MediaCodec( 6804): (0x7a962000) configure format: AMessage(what = 0x00000000) = {
D/MediaCodec( 6804):       Buffer csd-1 = {
D/MediaCodec( 6804):         00000000:  00 00 00 01 68 ef bc b0                           ....h...
D/MediaCodec( 6804):       }
D/MediaCodec( 6804):       string mime = "video/avc"
D/MediaCodec( 6804):       int32_t width = 1920
D/MediaCodec( 6804):       Buffer csd-0 = {
D/MediaCodec( 6804):         00000000:  00 00 00 01 67 64 00 28  ac d9 40 78 02 27 e5 c0  ....gd.(..@x.'..
D/MediaCodec( 6804):         00000010:  5a 80 80 80 a0 00 00 03  00 20 00 00 07 91 e3 06  Z........ ......
D/MediaCodec( 6804):         00000020:  32 c0                                             2.
D/MediaCodec( 6804):       }
D/MediaCodec( 6804):       int32_t height = 1080
D/MediaCodec( 6804):     }
I/ExtendedACodec( 6804): setupVideoDecoder()
I/ACodec  ( 6804): [OMX.qcom.video.decoder.avc] setupVideoDecoder Width Height (1920x1080)
I/ACodec  ( 6804): mime (video/avc) compressionFormat (7)
I/ExtendedACodec( 6804): Decoder will be in frame by frame mode
I/MediaCodec( 6804): (0x7a962000) start
I/MediaCodec( 6804): (0x7a962000) kWhatStartCompleted
E/ACodec  ( 6804): [OMX.qcom.video.decoder.avc] Failed to get port format to send format change
D/MediaCodec( 6804): (0x7a962000) kWhatOutputBuffersChanged
D/        ( 6804): PlayerBase::PlayerBase()
D/        ( 6804): TrackPlayerBase::TrackPlayerBase()
I/libOpenSLES( 6804): Emulating old channel mask behavior (ignoring positional mask 0x3, using default mask 0x3 based on channel count of 2)
I/AudioTrack( 6804): createTrack_l(0): AUDIO_OUTPUT_FLAG_FAST successful; frameCount 0 -> 1920
E/libOpenSLES( 6804): Configuration error: unknown key
W/libOpenSLES( 6804): Leaving AndroidConfiguration::GetConfiguration (SL_RESULT_PARAMETER_INVALID)
D/        ( 6804): PlayerBase::stop() from IPlayer
D/AudioTrack( 6804): stop(49): called with 0 frames delivered
I/media_kit( 6804): com.alexmercerind.mediakitandroidhelper.MediaKitAndroidHelper.deleteGlobalObjectRef: 10758

**Comment by NANI-SORE** [comment]
I have the same problem, but after setting video controller config like this videos started to work:
```
configuration: const VideoControllerConfiguration(
  vo: 'mediacodec_embed',
  hwdec: 'mediacodec',
),
```

Changing `androidAttachSurfaceAfterVideoParameters` to `false` also didn't work for me

**Comment by alexmercerind** [comment]
> I have the same problem, but after setting video controller config like this videos started to work:
> 
> ```
> configuration: const VideoControllerConfiguration(
>   vo: 'mediacodec_embed',
>   hwdec: 'mediacodec',
> ),
> ```

Very confusing, I will suggest against that however. I'm pretty sure you've read warnings about this configuration.

Does mpv for Android work well for your stream? A sample will help.

**Comment by NANI-SORE** [comment]
Yeah, I understand, I just used it as a temporary solution while I develop. I want to use this package as an alternative video backend and will give users settings to select different options for vo/hwdec just in case defaults won't work for them.

I tried again today and videos actually started working with default values (vo: gpu, hwdec: auto-safe) on device where it was failing before, but I didn't change anything 🤔.

Still happens on an emulator (Pixel 3a API 34) on example/test app even with default videos

**Comment by alexmercerind** [comment]
Please don't use emulator as a reference. It's very buggy, I also get a black screen on my M1 Air.

We also enforce software decoding in iOS simulators & Android emulators. It does not give idea about actual performance & stability.

**Comment by NANI-SORE** [comment]
Okay, guess I won't be using emulators for this. I'll report if it'll happen again on any of my physical devices

**Comment by adilmehmood-md** [comment]
> I have the same problem, but after setting video controller config like this videos started to work:
> 
> ```
> configuration: const VideoControllerConfiguration(
>   vo: 'mediacodec_embed',
>   hwdec: 'mediacodec',
> ),
> ```
> 
> Changing `androidAttachSurfaceAfterVideoParameters` to `false` also didn't work for me

Can you mention the plugin version you are using? i am not getting the option `hwdec` in VideoControllerConfiguration as it's replaced with `PlayerConfiguration` now.

**Comment by alexmercerind** [comment]
@adilmehmood-md,

DO NOT DO IT.

- Try `androidAttachSurfaceAfterVideoParameters` first.
- `VideoControllerConfiguration` is passed to `VideoController` (not `Player`).

**Comment by alexmercerind** [comment]
Also don't use emulator.

**Comment by alexmercerind** [comment]
VideoControllerConfiguration

**Comment by Feichtmeier** [comment]
I face the same problem with using it on linux and macos
there are german tv stations using m3u8 with hls
the video is opened but then nothing. no error, no stream

**Comment by wiedymi** [comment]
I have same issue, but with dash stream

**Comment by 2shrestha22** [comment]
work fine for first play but when i play another video it is black with audio only. Impeller is not enabled.

**Comment by krishanroy** [comment]
@alexmercerind I am also facing the same issue while implementing `media-kit` into my work app to play a HLS .m3u8 file. It works fine on iOS, but on Android I am only getting audio with a black screen. First once or twice it works fine but then just audio with a black screen.

**Comment by dev01prishusoft** [comment]
Is there any solution i only need to work in android

**Comment by ust1957** [comment]
docs need some update with hints or example to use these common settings .

**Comment by yisar** [comment]
俺也一样
